server/src-lib/Hasura/RQL/Types/Common.hs

```diff
@@ -182,6 +192,61 @@ instance Hashable ForeignKey
 
 newtype FunctionArgName =
   FunctionArgName { getFuncArgNameTxt :: T.Text}
-  deriving (Show, Eq, ToJSON)
+  deriving (Show, Eq, ToJSON, FromJSON, Lift, DQuote, IsString)
 
 type CustomColumnNames = HM.HashMap PGCol G.Name
+
+data FunctionArg
+  = FunctionArg
+  { faName       :: !(Maybe FunctionArgName)
+  , faType       :: !QualifiedPGType
+  , faHasDefault :: !Bool
+  } deriving (Show, Eq)
+$(deriveToJSON (aesonDrop 2 snakeCase) ''FunctionArg)
+
+newtype ComputedFieldName =
+  ComputedFieldName { unComputedFieldName :: NonEmptyText}
+  deriving (Show, Eq, Lift, FromJSON, ToJSON, Q.ToPrepArg, DQuote, Hashable)
+
+computedFieldNameToText :: ComputedFieldName -> Text
+computedFieldNameToText = unNonEmptyText . unComputedFieldName
+
+fromComputedField :: ComputedFieldName -> FieldName
+fromComputedField = FieldName . computedFieldNameToText
+
+data FunctionTableArgument
+  = FTAFirstArgument
+  | FTAName !FunctionArgName
+  deriving (Show, Eq)
+
+instance ToJSON FunctionTableArgument where
+  toJSON FTAFirstArgument  = String "first_argument"
+  toJSON (FTAName argName) = object ["name" .= argName]
+
+data ComputedFieldReturn
+  = CFRScalar !PGScalarType
+  | CFRSetofTable !QualifiedTable
+  deriving (Show, Eq)
+$(deriveToJSON defaultOptions { constructorTagModifier = snakeCase . drop 3
+                              , sumEncoding = TaggedObject "type" "info"
+                              }
+   ''ComputedFieldReturn
+ )
+
+data ComputedFieldFunction
+  = ComputedFieldFunction
+  { _cffName          :: !QualifiedFunction
+  , _cffInputArgs     :: !(Seq.Seq FunctionArg)
+  , _cffTableArgument :: !FunctionTableArgument
+  , _cffDescription   :: !(Maybe PGDescription)
+  } deriving (Show, Eq)
+$(deriveToJSON (aesonDrop 4 snakeCase) ''ComputedFieldFunction)
+
+data ComputedFieldInfo
+  = ComputedFieldInfo
+  { _cfiName       :: !ComputedFieldName
+  , _cfiFunction   :: !ComputedFieldFunction
+  , _cfiReturnType :: !ComputedFieldReturn
+  , _cfiComment    :: !(Maybe T.Text)
+  } deriving (Show, Eq)
+$(deriveToJSON (aesonDrop 4 snakeCase) ''ComputedFieldInfo)
```

Comment:
Could these types perhaps go in their own module, maybe `Hasura.RQL.Types.ComputedField`? It’d be nice to avoid making `Common.hs` too big if we can help it, and it doesn’t look to me like anything else in this module depends on any of these types (but I could be missing something).